allennlp/models/coreference_resolution/coref.py

```diff
@@ -0,0 +1,707 @@
+        # Shape: (batch_size, num_spans_to_keep, max_antecedents, embedding_size)
+        antecedent_distance_embeddings = antecedent_distance_embeddings.expand(*expanded_distance_embeddings_shape)
+
+        # Shape: (batch_size, num_spans_to_keep, max_antecedents, embedding_size)
+        pairwise_embeddings = torch.cat([target_embeddings,
+                                         antecedent_embeddings,
+                                         similarity_embeddings,
+                                         antecedent_distance_embeddings], -1)
+        return pairwise_embeddings
+
+    @staticmethod
+    def _compute_antecedent_gold_labels(top_span_labels: torch.IntTensor,
+                                        antecedent_labels: torch.IntTensor):
+        """
+        Parameters
+        ----------
+        top_span_labels: torch.IntTensor
+            The cluster id label for every span.
+        antecedent_labels: torch.IntTensor
+            The cluster id label for every antecedent span.
+        Returns
+        -------
+        augmented_label: torch.FloatTensor
+            A binary indicator for every pair of spans. This label is one if and only if ithe
+            pair of spans belong to the same cluster. The labels are augmented with a dummy
+            antecedent at the zeroth position, which represents the prediction that a span does
+            not have any antecedent.
+        """
+        # Shape: (batch_size, num_spans_to_keep, max_antecedents)
+        target_labels = top_span_labels.expand_as(antecedent_labels)
+        same_cluster_indicator = (target_labels == antecedent_labels).float()
+        non_dummy_indicator = (target_labels >= 0).float()
+        pairwise_labels = same_cluster_indicator * non_dummy_indicator
+
+        # Shape: (batch_size, num_spans_to_keep, 1)
+        dummy_labels = (1 - pairwise_labels).prod(-1, keepdim=True)
+
+        # Shape: (batch_size, num_spans_to_keep, max_antecedents + 1)
+        augmented_labels = torch.cat([dummy_labels, pairwise_labels], -1)
+        return augmented_labels
+
+    @staticmethod
+    def _compute_negative_marginal_log_likelihood(augmented_antecedent_scores: torch.FloatTensor,
+                                                  augmented_labels: torch.IntTensor,
+                                                  top_span_mask: torch.FloatTensor):
+        """
+        Parameters
+        ----------
+        augmented_antecedent_scores: torch.FloatTensor
+            The pairwise between every span and its possible antecedents.
+        augmented_labels: torch.IntTensor
+            A binary indicator for whether it is consistent with the data for a span to choose
+            an antecedent.
+        Returns
+        -------
+        negative_marginal_log_likelihood: torch.FloatTensor
+            The negative marginal loglikelihood of the gold cluster labels. This computes the
```

Comment:
I think for a lot of these methods you've put the description of the logic inside the description of the return value.  I think it makes more sense to put the description of the logic at the top, then give more detail as necessary in the description of each parameter and return value (or just give a name and type here, if the description above is already good enough).